### PR TITLE
data model: transform marc21 field containing identifiers

### PR DESCRIPTION
**Pull request by reropag**
Only transformation code and unit test for identifiers are included in this PR.
No testing can be done at this stage.

### CODE REVIEW

**Comment by jma** [comment]
change data_model into data model in the commit message

File: rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py [hunk 1]
```diff
@@ -358,31 +358,281 @@ def marc21_to_abstracts(self, key, value):
...
 @utils.ignore_value
-def marc21_to_identifier_isbn(self, key, value):
-    """Get identifier isbn.
+def marc21_to_identifiedBy_from_field_020(self, key, value):
+    """Get identifier from field 020."""
+    def populate_acquisitionsTerms_qualifier(idendifier):
+        subfield_c = value.get('c')
+        if subfield_c:
+            subfield_c = subfield_c.strip()
+            idendifier['acquisitionsTerms'] = subfield_c
```
Comment (jma) [inline rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py]: -> identifier

File: rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py [hunk 2]
```diff
@@ -358,31 +358,281 @@ def marc21_to_abstracts(self, key, value):
...
+        subfields_q = value.get('q')
```
Comment (jma) [inline rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py]:
```python
        subfields_q = value.get('q')
        if subfields_q:
            for qualifier in utils.force_list(subfields_q):
                qualifier = qualifier.strip()
                if 'qualifier' in idendifier:
                    idendifier['qualifier'] += ', ' + qualifier
                else:
                    idendifier['qualifier'] = qualifier
```
can be replaced by
```python
if value.get('q'):
    idendifier['qualifier'] = ', '.join(utils.force_list(value.get('q'))
```

File: rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py [hunk 3]
```diff
@@ -358,31 +358,281 @@ def marc21_to_abstracts(self, key, value):
...
+                    idendifier['qualifier'] += ', ' + qualifier
+                else:
+                    idendifier['qualifier'] = qualifier
+    idendifier_added = 0
+    identifiedBy = self.get('identifiedBy', [])
+
+    def build_identifier_from(subfield_data, status=None):
+        subfield_data = subfield_data.strip()
+        idendifier = {}
+        idendifier['value'] = subfield_data
```
Comment (jma) [inline rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py]: can be merge with the previous line

File: rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py [hunk 4]
```diff
@@ -358,31 +358,281 @@ def marc21_to_abstracts(self, key, value):
...
+        build_identifier_from(subfield_a)
+        idendifier_added = 1
+    subfield_z = value.get('z')
+    if subfield_z:
+        build_identifier_from(subfield_z, status='invalid or cancelled')
+        idendifier_added = 1
+    if idendifier_added:
+        return identifiedBy
+    else:
+        return None
```
Comment (jma) [inline rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py]: `return identifiedBy or None`

File: rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py [hunk 5]
```diff
@@ -358,31 +358,281 @@ def marc21_to_abstracts(self, key, value):
...
-        identifiers = self.get('identifiers', {})
-        identifiers['isbn'] = value.get('a')
-        return identifiers
+
+@marc21tojson.over('identifiedBy', '^022..')
+@utils.ignore_value
+def marc21_to_identifiedBy_from_field_022(self, key, value):
+    """Get identifier from field 022."""
+    status_for = {}
+    status_for['m'] = 'cancelled'
```
Comment (jma) [inline rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py]:
```python
    status_for = {}
    status_for['m'] = 'cancelled'
    status_for['y'] = 'invalid'

    type_for = {}
    type_for['a'] = 'bf:Issn'
    type_for['l'] = 'bf:IssnL'
    type_for['m'] = 'bf:IssnL'
    type_for['y'] = 'bf:Issn'
```
replaced by
```python
    status_for = {
        'm': 'cancelled',
        'y': 'invalid'
    }
    ...
```

File: rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py [hunk 6]
```diff
@@ -358,31 +358,281 @@ def marc21_to_abstracts(self, key, value):
...
+    status_for['y'] = 'invalid'
+
+    type_for = {}
+    type_for['a'] = 'bf:Issn'
+    type_for['l'] = 'bf:IssnL'
+    type_for['m'] = 'bf:IssnL'
+    type_for['y'] = 'bf:Issn'
+
+    idendifier_added = 0
```
Comment (jma) [inline rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py]: same as before

File: rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py [hunk 7]
```diff
@@ -358,31 +358,281 @@ def marc21_to_abstracts(self, key, value):
...
+        if subfield_c:
+            subfield_c = subfield_c.strip()
+            idendifier['acquisitionsTerms'] = subfield_c
+        subfield_d = value.get('d')
+        if subfield_d:
+            subfield_d = subfield_d.strip()
+            idendifier['note'] = subfield_d
+        subfields_q = value.get('q')
+        if subfields_q:
+            for qualifier in utils.force_list(subfields_q):
```
Comment (jma) [inline rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py]: can be replaced by a `split`

File: rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py [hunk 8]
```diff
@@ -358,31 +358,281 @@ def marc21_to_abstracts(self, key, value):
...
+                    idendifier['qualifier'] += ', ' + qualifier
+                else:
+                    idendifier['qualifier'] = qualifier
+
+    idendifier = {}
+    subfield_a = value.get('a')
+    subfield_2 = value.get('2')
+    if subfield_a:
+        subfield_a = subfield_a.strip()
+        if re.search('permalink\.snl\.ch', subfield_a):
```
Comment (jma) [inline rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py]: please use `r''` for regex

File: rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py [hunk 9]
```diff
@@ -358,31 +358,281 @@ def marc21_to_abstracts(self, key, value):
...
+            idendifier['value'] = subfield_a
```
Comment (jma) [inline rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py]: `identifier.update({'value': subfield_a, 'type': 'uri' ....})`

File: rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py [hunk 10]
```diff
@@ -358,31 +358,281 @@ def marc21_to_abstracts(self, key, value):
...
+            idendifier['source'] = 'SNL'
+        elif re.search('bnf\.fr/ark', subfield_a):
+            idendifier['value'] = subfield_a
+            idendifier['type'] = 'uri'
+            idendifier['source'] = 'BNF'
+        elif subfield_2:
+            subfield_2 = subfield_2.strip()
+            idendifier['value'] = subfield_a
+            populate_acquisitionsTerms_note_qualifier(idendifier)
+            if re.search('doi', subfield_2, re.IGNORECASE):
```
Comment (jma) [inline rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py]: dict with regex

File: rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py [hunk 11]
```diff
@@ -358,31 +358,281 @@ def marc21_to_abstracts(self, key, value):
...
+                elif indicateur_1 == '3':
+                    match = re.search(r'^97', subfield_a)
+                    if match:  # $a starts with '97'
+                        idendifier['type'] = 'bf:Ean'
+                    else:
+                        idendifier['type'] = 'bf:Identifier'
+                elif indicateur_1 == '8':
+                    # 33 chars example: 0000-0002-A3B1-0000-0-0000-0000-2
+                    pattern = r'^(.{24}|.{26}|(.{4}-){4}.-(.{4}\-){2}.)$'
+                    prog = re.compile(pattern)
```
Comment (jma) [inline rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py]: compile is useless

File: rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py [hunk 12]
```diff
@@ -358,31 +358,281 @@ def marc21_to_abstracts(self, key, value):
...
-    identifiers['reroID'] = value.get('a')
-    return identifiers
+
+@marc21tojson.over('identifiedBy', '^028..')
+@utils.ignore_value
+def marc21_to_identifiedBy_from_field_028(self, key, value):
+    """Get identifier from field 028."""
+    def populate_qualifier():
+        subfields_q = value.get('q')
+        if subfields_q:
```
Comment (jma) [inline rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py]: if is useless can be rewritten as previous

File: rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py [hunk 13]
```diff
@@ -358,31 +358,281 @@ def marc21_to_abstracts(self, key, value):
...
+            for qualifier in utils.force_list(subfields_q):
+                qualifier = qualifier.strip()
+                if 'qualifier' in idendifier:
+                    idendifier['qualifier'] += ', ' + qualifier
+                else:
+                    idendifier['qualifier'] = qualifier
+
+    type_for_ind1 = {}
```
Comment (jma) [inline rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py]: can be defined in the dict creation

File: rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py [hunk 14]
```diff
@@ -358,31 +358,281 @@ def marc21_to_abstracts(self, key, value):
...
+        if subfield_b:
+            idendifier['source'] = subfield_b.strip()
+        indicateur_1 = key[3]
+        if type_for_ind1[indicateur_1]:
+            idendifier['type'] = type_for_ind1[indicateur_1]
+        else:
+            idendifier['type'] = 'bf:Identifier'
+        identifiedBy = self.get('identifiedBy', [])
+        identifiedBy.append(idendifier)
+        return identifiedBy
```
Comment (jma) [inline rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py]: you can use `or`

File: rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py [hunk 15]
```diff
@@ -358,31 +358,281 @@ def marc21_to_abstracts(self, key, value):
...
+
+@marc21tojson.over('identifiedBy', '^035..')
+@utils.ignore_value
+def marc21_to_identifiedBy_from_field_035(self, key, value):
+    """Get identifier from field 035."""
+    subfield_a = value.get('a')
+    if subfield_a:
+        subfield_a = subfield_a.strip()
+        idendifier = {}
+        idendifier['value'] = subfield_a
```
Comment (jma) [inline rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py]: can be done in the dict definition

File: rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py [hunk 16]
```diff
@@ -358,31 +358,281 @@ def marc21_to_abstracts(self, key, value):
...
+            # match.group(1) : parentheses content
+            idendifier['source'] = match.group(1)
+            # value without parenthesis and parentheses content
+            idendifier['value'] = match.group(2)
+        else:
+            idendifier['value'] = subfield_a
+        idendifier['type'] = 'bf:Local'
+        identifiedBy = self.get('identifiedBy', [])
+        identifiedBy.append(idendifier)
+        return identifiedBy
```
Comment (jma) [inline rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py]: please use `or`

File: tests/unit/test_documents_dojson.py [hunk 17]
```diff
@@ -605,3 +582,373 @@ def test_marc21_to_subjects():
...
+        {
+            'type': 'bf:IssnL',
+            'status': 'cancelled',
+            'value': '0080-4650'
+        }
+    ]
+
+
+def test_marc21_to_identifiedBy_from_024_snl_bnf():
+    """Test dojson identifiedBy from 024 field sbn and bnf."""
```
Comment (jma) [inline tests/unit/test_documents_dojson.py]: sbn -> isbn?

File: tests/unit/test_documents_dojson.py [hunk 18]
```diff
@@ -605,3 +582,373 @@ def test_marc21_to_subjects():
     marc21json = create_record(marc21xml)
     data = marc21tojson.do(marc21json)
     assert data.get('subjects') == ['subjects 600', 'subjects 666']
+
+
```
Comment (rerowep) [inline tests/unit/test_documents_dojson.py]: Repetitive tags not tested!

File: rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py [hunk 19]
```diff
@@ -358,31 +358,255 @@ def marc21_to_abstracts(self, key, value):
...
+            # value without parenthesis and parentheses content
+            identifier['value'] = match.group(1)
+        if status:
+            identifier['status'] = status
+        identifier['type'] = 'bf:Isbn'
+        identifiedBy.append(identifier)
+
+    identifiedBy = self.get('identifiedBy', [])
+    subfield_a = value.get('a')
+    if subfield_a:
```
Comment (rerowep) [inline rero_ils/modules/documents/dojson/contrib/marc21tojson/model.py]: Maybe we have multiple 020:
```
    subfields_a = utils.force_list(value.get('a'))
    for subfield_a in subfields_a:
        build_identifier_from(subfield_a)
        subfield_z = value.get('z')
        if subfield_z:
            build_identifier_from(subfield_z, status='invalid or cancelled')
```

**Pull request merged**